loader/config.c

```diff
@@ -5,878 +5,876 @@
+        cfg_ungetch(parser);
 
-            if (IS(STATE_VALUE)) {
-                if (!finalize_key_value(cfg, s, false))
-                    return false;
+        token->ident.text = &parser->source[parser->pos.pos];
+        token->ident.size = 0;
 
-                set_state(s, STATE_NORMAL);
-                continue;
-            }
+        while (isalnum(ch) || ch == '_' || ch == '-') {
+            ch = cfg_getch(parser);
+            token->ident.size++;
+        }
 
-            if (IS(STATE_COMMENT)) {
-                set_state(s, STATE_NORMAL);
-                continue;
-            }
+        token->ident.size--;
+        cfg_ungetch(parser);
+
+        if (sv_equals(token->ident, SV("null"))) {
+            token->type = TK_NULL;
+        } else if (sv_equals(token->ident, SV("true"))) {
+            token->type    = TK_BOOL;
+            token->is_true = true;
+        } else if (sv_equals(token->ident, SV("false"))) {
+            token->type    = TK_BOOL;
+            token->is_true = false;
+        }
 
-            PARSE_ERROR("invalid character");
+        return true;
+      }
+    case '\r':
+    case '\n':
+        return cfg_fetch_token(parser, token);
+    default:
+        return cfg_raise(parser, &token->pos, "Unknown character");
+    }
+}
 
-        case '=':
-            if (IS(STATE_NORMAL) || (IS(STATE_VALUE) && !s->open_quote_character))
-                PARSE_ERROR("invalid character");
+static struct config_entry *cfg_fetch_entry(struct config_parser *parser)
+{
+    struct config_entry *ent = parser->cfg->size >= parser->cfg->capacity ? NULL :
+                              &parser->cfg->buffer[parser->cfg->size++];
 
-            if (IS(STATE_KEY)) {
-                set_state(s, STATE_VALUE);
-                continue;
-            }
+    if (ent == NULL) {
+        size_t cap = MAX(parser->cfg->capacity * 2, PAGE_SIZE / sizeof(struct config_entry));
 
-            sv_extend_by(&s->current_value_view, 1);
-            continue;
+        if (!(ent = allocate_bytes(sizeof(struct config_entry) * cap)))
+            return NULL;
 
-        case ':':
-            if (IS(STATE_NORMAL))
-                PARSE_ERROR("invalid character");
+        memcpy(ent, parser->cfg->buffer, sizeof(struct config_entry) * parser->cfg->size);
 
-            if (IS(STATE_KEY)) {
-                if (!finalize_key_value(cfg, s, true))
-                    return false;
+        parser->cfg->capacity = cap;
+        parser->cfg->buffer   = ent;
 
-                set_state(s, STATE_NORMAL);
-                s->expecting_end_of_value = true;
-                continue;
-            }
+        return cfg_fetch_entry(parser);
+    }
 
-            if (IS(STATE_VALUE) && !s->open_quote_character)
-                PARSE_ERROR("invalid character");
+    return ent;
+}
 
-            sv_extend_by(&s->current_value_view, 1);
-            continue;
+static ssize_t cfg_parse_object(struct config_parser *parser, size_t lvl)
```

Comment:
This function's return values are really hard to interpret without going through the code. Could you maybe write a comment above explaining how different return values should be interpreted?
E.g:
```
/*
 * Returns -1 in case of an error, otherwise returns ...
 * ... more explanation
 */

Reply:
Or maybe even make a macro like `RETURN_CHILD_COUNT(x) (-(x) - 2)`?